### PR TITLE
Add KNX tools: telegram history + entity read/write

### PR DESCRIPTION
**Pull request by pos-ei-don**
## What

Adds a small **KNX tool set** to the MCP server, mirroring Home Assistant's KNX websocket commands via `hass.data[KNX_MODULE_KEY]`:

**Read**
- `knx_recent_telegrams` — recent group-monitor telegram history (incl. **source device** + decoded value), regex-filterable. Retrospective: find which KNX device wrote a given group address.
- `knx_get_base_data` — connection status, gateway address, xknx version, loaded ETS project info, UI-creatable platforms.
- `knx_get_entities` — group-address ↔ entity bindings (the KNX-specific view generic entity tools don't expose), regex filter on GA.

**Write (experimental)** — mutate the KNX UI config via `config_store`:
- `knx_create_entity` (`platform` + `data` → entity_id), `knx_update_entity`, `knx_delete_entity`.

## Why

The server already writes KNX via `call_service`→`knx.send` and reads entity state, but there's no way to inspect the **raw KNX bus**, the **group-address↔entity bindings**, or to **create/manage KNX UI entities** programmatically. These tools close that gap (e.g. diagnosing a value that flaps at dusk, or scripting missing KNX objects into HA).

## Notes

- All KNX-module access is **defensive**: `KNX_MODULE_KEY` import is guarded, and each tool returns a friendly "KNX is not set up" / "unavailable on this HA version" message instead of raising — the server still loads cleanly without the KNX integration.
- Write tools are marked **experimental** (consistent with the existing helper write-tools).
- Unit tests in `tests/test_tools_knx.py`; tool-count test bumped to 65; README updated with a **KNX** section. Verified green on the test + integration jobs across HA 2025.4 / 2025.10 / 2026.4.

### CODE REVIEW

File: custom_components/mcp_server_http_transport/tools/knx.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        for t in telegrams
+        if (ga_re is None or ga_re.search(_destination(t)))
+        and (name_re is None or name_re.search(str(t.get("destination_name", ""))))
+    ]
+
+    limit = arguments.get("limit") or 200
+    returned = matched[-limit:]
+    timestamps = [t.get("timestamp") for t in telegrams if t.get("timestamp")]
+
+    return {
```
Comment (ganhammar) [inline custom_components/mcp_server_http_transport/tools/knx.py]: Every other tool handler returns the `{"content": [{"type": "text", "text": ...}]}` envelope with the payload run through `_HAJSONEncoder`, but these success paths return raw dicts. Two snags: MCP clients expect `result.content`, and the final response is serialized by aiohttp's default encoder rather than `_HAJSONEncoder`, so a real telegram/project object that isn't plain-primitive will 500 the whole response. The error paths here already do it right, only the success ones drifted. Tests don't catch it because they call the handler directly and skip HTTP serialization.

Same applies to the success returns in `knx_get_base_data`, `knx_get_entities`, and the three write tools.

File: custom_components/mcp_server_http_transport/tools/knx.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        return {"content": [{"type": "text", "text": f"Invalid regex: {err}"}]}
+
+    matched = [
+        t
+        for t in telegrams
+        if (ga_re is None or ga_re.search(_destination(t)))
+        and (name_re is None or name_re.search(str(t.get("destination_name", ""))))
+    ]
+
+    limit = arguments.get("limit") or 200
```
Comment (ganhammar) [inline custom_components/mcp_server_http_transport/tools/knx.py]: `arguments.get("limit") or 200` quietly turns an explicit `limit: 0` into 200, and a negative limit slices from the wrong end. Minor, but `max(1, int(arguments.get("limit") or 200))` would keep it predictable. Same line shows up in `knx_get_entities` too.

**Review by ganhammar — CHANGES_REQUESTED** [review]
Thanks for this, really nice addition. The KNX coverage is genuinely useful and the defensive guarding for when KNX isn't installed is well done.

One thing to sort before merge: the success paths return raw dicts instead of the `{"content": [...]}` envelope the other tools use, which both breaks the MCP result shape and skips `_HAJSONEncoder` on the way out (details inline). The error paths already do it right, so it's a small change. Left a minor note on `limit` handling too.

Tests, README, and tool-count bumps all look great. Thanks again!

**Comment by pos-ei-don** [comment]
Thanks for the detailed review! Here's what changed in this update:

**Content envelope** — all six success-path returns in `knx.py` now use the standard MCP envelope (`{"content": [{"type": "text", "text": json.dumps(..., cls=_HAJSONEncoder)}]}`). I also added `import json` and pulled in `_HAJSONEncoder` from the tools `__init__`.

**limit clamping** — the `or 200` fallback silently converts `limit=0` to 200 (because `0` is falsy). Fixed at both call sites with an explicit `None` check:
```python
_lim = arguments.get("limit")
limit = max(1, int(_lim) if _lim is not None else 200)
```

**Tests** — updated all success-path assertions to unpack the envelope via a small `_unpack()` helper, and added three new edge-case tests: `limit=0` and `limit<0` for both `knx_recent_telegrams` and `knx_get_entities`. All 18 tests pass.

**Comment by pos-ei-don** [comment]
## Live HA Test Results ✅

Tested on live HA 2026.6.1 with real KNX bus (xknx 3.15.0, individual address 1.0.3, ETS project loaded):

### knx_get_base_data
Returns KNX connection status, gateway address, xknx version, ETS project metadata — working correctly with MCP content envelope.

### knx_recent_telegrams
- Buffer: 25,000 telegrams (~7.5 hours of bus traffic)
- `limit=5` → returns 5 most-recent ✅
- `limit=0` → correctly clamped to 1 (fix verified on live instance) ✅
- `filter_ga="^1/1/1"` → 437 matching, returns requested count ✅

### knx_get_entities
Returns 794 KNX group addresses with bound entity identifiers — working correctly.

---

### Additional fix in this PR

`manifest.json`: removed `"logbook"` from `after_dependencies`. `logbook` was removed from HA in 2025.x; keeping it there caused a startup cascade failure (`recorder` timing → `logbook` → `mcp_server_http_transport` fails to load entirely) on HA 2025.x/2026.x.

**Before:** `"after_dependencies": ["recorder", "lovelace", "logbook", "camera"]`  
**After:** `"after_dependencies": ["recorder", "lovelace", "camera"]`

**Comment by ganhammar** [comment]
Thanks for the quick turnaround! The envelope and `limit` fixes look great, verified both. 🙌

Two things on the manifest before this is mergeable, and the failing **Validate** check (hassfest) points right at them:

```
[ERROR] Using component knx but it's not in 'dependencies' or 'after_dependencies'
[ERROR] Using component logbook but it's not in 'dependencies' or 'after_dependencies'
```

- `knx` needs adding to `after_dependencies` since the new tools import from it.
- On `logbook`: it actually [still exists in HA core](https://www.home-assistant.io/integrations/logbook/) (it's in current `dev`, docs page is live), and removing it is what's tripping the second error, so something still references it. Could you double-check the claim that it was removed? If you hit a real startup issue, I'd love the log so we can fix the actual cause, but I think this hunk should come out.

Everything else looks solid. Thanks again!

**Comment by pos-ei-don** [comment]
Thanks for the review! Both points are valid — fixed in 4793594.

- **`knx` added to `after_dependencies`** — correct catch, the defensive `try/except` import doesn't exempt it from hassfest's dependency check.
- **`logbook` restored** — you're right, I was wrong to remove it. The original removal was based on an incorrect assumption that it had been dropped from core; I didn't verify against current `dev`. It's back.

`after_dependencies` is now: `["recorder", "lovelace", "logbook", "camera", "knx"]`

**Review by ganhammar — APPROVED** [review]
Looks great now! Thanks for your contribution!
